pennylane/labs/dla/structure_constants_dense.py

```diff
@@ -0,0 +1,122 @@
+    :class:`~PauliSentence` that are employed in :func:`~structure_constants`, e.g., when there are few generators
+    that are sums of many Paulis.
+
+    .. seealso:: For details on the mathematical definitions, see :func:`~structure_constants` and the section "Lie algebra basics" in our `g-sim demo <https://pennylane.ai/qml/demos/tutorial_liesim/#lie-algebra-basics>`__.
+
+    Args:
+        g (np.array): The (dynamical) Lie algebra provided as dense matrices, as generated from :func:`pennylane.labs.lie.lie_closure_dense`.
+            ``g`` should have shape ``(d, 2**n, 2**n)`` where ``d`` is the dimension of the algebra and ``n`` is the number of qubits. Each matrix ``g[i]`` should be Hermitian.
+        is_orthonormal (bool): Whether or not the matrices in ``g`` are orthonormal with respect to the Hilbert-Schmidt inner product on
+            (skew-)Hermitian matrices. If the inputs are orthonormal, it is recommended to set ``is_orthonormal`` to ``True`` to reduce
+            computational cost. Defaults to ``True``.
+
+    Returns:
+        TensorLike: The adjoint representation of shape ``(d, d, d)``, corresponding to
+        the indices ``(gamma, alpha, beta)``.
+
+    **Example**
+
+    Let us generate the DLA of the transverse field Ising model using :func:`~lie_closure_dense`.
+
+    >>> n = 4
+    >>> gens = [X(i) @ X(i+1) + Y(i) @ Y(i+1) + Z(i) @ Z(i+1) for i in range(n-1)]
+    >>> g = lie_closure_dense(gens)
+    >>> g.shape
+    (12, 16, 16)
+
+    The DLA is represented by a collection of twelve :math:`2^4 \times 2^4` matrices.
+    Hence, the dimension of the DLA is :math:`d = 12` and the structure constants have shape ``(12, 12, 12)``.
+
+    >>> adj = structure_constants_dense(g)
+    >>> adj.shape
+    (12, 12, 12)
+
+    **Internal representation**
+
+    As mentioned above, the input is assumed to be a batch of Hermitian matrices, even though
+    algebra elements are usual skew-Hermitian. That is, the input should represent the operators
+    :math:`G_\alpha` for an algebra basis :math:`\{iG_\alpha\}_\alpha`.
+    In an orthonormal basis of this form, the structure constants can then be computed simply via
+
+    .. math::
+
+        f^\gamma_{\alpha, \beta} = \text{tr}[-i G_\gamma[iG_\alpha, iG_\beta]] = i\text{tr}[G_\gamma [G_\alpha, G_\beta]] \in \mathbb{R}.
+
+    **Structure constants in non-orthonormal bases**
+
+    Structure constants are often discussed using an orthonormal basis of the algebra.
+    This function can deal with non-orthonormal bases as well. For this, the Gram
+    matrix :math:`g` between the basis elements is taken into account when computing the overlap
+    of a commutator `:math:`[iG_\alpha, iG_\beta]` with all algebra elements :math:`iG_\gamma`.
+    The resulting formula reads
+
+    .. math::
+
+        f^\gamma_{\alpha, \beta} &= \sum_\eta g^{-1}_{\gamma\eta} i \text{tr}[G_\eta [G_\alpha, G_\beta]]\\
+        g_{\gamma \eta} &= \text{tr}[G_\gamma G_\eta] \quad(\in\mathbb{R})
```

Comment:
As an ex-researcher who worked on General Relativity, my heart suffers when I see covariant and contravariant tensor indices mixed on two sides of an equation. But In this case it's fine since structure constants are real numbers, and the dual space of R is canonically isomorphic to R 🚀